content/blog/2023/2023-11-23-sorting-basics-kotlin.md

```diff
@@ -0,0 +1,108 @@
+---
+title: "Guide to Sorting in Kotlin"
+categories: ["Kotlin"]
+date: 2023-11-23 00:00:00 +1100
+authors: [ezra]
+excerpt: "In this tutorial, we'll discuss what sorting is and discuss the various methods we can use to sort various elements."
+image: images/stock/0104-on-off-1200x628-branded.jpg
+url: sorting in kotlin
+---
+
+Sorting refers to the process of arranging elements in a specific order. The order could be ascending or descending based on certain criteria such as numerical or lexicographical order. In this guide, we'll explore various sorting techniques and functions available in Kotlin, unveiling the simplicity and flexibility the Kotlin language offers.
+
+## Basic Sorting in Kotlin
+
+Let's take a look at how we can sort lists in Kotlin:
+
+```kotlin
+fun main() {
+    val numbers = listOf(4, 2, 8, 1, 5)
+
+    val sortedNumbers = numbers.sorted()
+    println("Sorted numbers: $sortedNumbers")
+
+    val descendingNumbers = numbers.sortedDescending()
+    println("Descending numbers: $descendingNumbers")
+}
+```
+
+In this code, we first create a list of `numbers`, sort them in ascending order using `sorted()` and then sort them in descending order using `sortedDescending()`.
+
+Here is another example of how we can sort arrays in Kotlin:
+
+```kotlin
+fun main() {
+    val numbersArray = arrayOf(4, 2, 8, 1, 5)
+
+    val sortedNumbers = numbersArray.sorted()
+    println("Sorted numbers: $sortedNumbers")
+
+    val descendingNumbers = numbersArray.sortedDescending()
+    println("Descending numbers: $descendingNumbers")
+}
+
+```
+
+In our code above, we initialize an array named `numbersArrays` with integers and demonstrate sorting operations. First, it employs the sorted() function to create a new array `sortedArrayNumbers` containing the elements of numbers in ascending order. The sorted array is then printed to the console using contentToString(). Subsequently, the sortedDescending() function is applied to obtain another array, descendingNumbers, with the elements sorted in descending order and this array is printed as well.
+
+
+## SortBy
+
+The sortBy function in Kotlin is used to sort a collection for example a list or an array based on a specified key or custom sorting criteria.
+
+ Here's an example using sortBy:
+
+ ```kotlin
+ data class Person(val name: String, val position: Int)
+
+fun main() {
+    val people = listOf(
+        Person("John", 1),
+        Person("Doe", 2),
+        Person("Mary",3)
+    )
+
+    // Sorting people by position in ascending order
+    val sortedPeople = people.sortedBy { it.position }
+    println("Sorted by position: $sortedPeople")
+
+    // Sorting people by position in descending order
+    val reverseSortedPeople = people.sortedByDescending { it.position }
+    println("Reverse sorted by position: $reverseSortedPeople")
```

Comment:
Another example that shows a bit more complicated sorting like "first sort by position, then sort by name" would be good to show that you can put more than just `it.position` into the `{}` ...